### PR TITLE
deliver #825 add Map::deleteAll for removing multiple keys from a map

### PR DESCRIPTION
**Pull request by vinnymac**
#### Summary

After some thought and benchmarking, `withMutations` turned out to be a performant way of implementing `deleteAll`. I tested against some other implementations such as mapping the keys to an object and merging, creating a new map for each key (obviously slow), and filtering the map by keys that did not exist in the provided iterable. They all resulted in slower benchmarks.
#### Questions
1. I see `iter`, `k`, and `v` being used throughout the codebase, while the contributions state to not use abbreviations, I avoided them wherever possible, is this a rule I should be following?
2. Should I be using `forceIterator` on the `keys` parameter or is my usage of `IndexedIterable` applicable here?
3. In `withMutations` I use `remove`, is it okay even though it is an [undocumented behavior](https://github.com/facebook/immutable-js/issues/228#issuecomment-67554984)? It works correctly for `Map` in this case.
4. I have never worked with `TypeScript` or `Flow` before, did I identify the supported types properly?
5. Should we keep the behavior of returning the current `Map` when no `keys` are passed in or assert that the `keys` parameter is defined and throw if it is not defined? i.e. `Map().deleteAll() // The required parameter keys was not provided.`
6. Do we want to also provide the alias `subtract` for this method on `Map`?
#### ToDo
- [x] Add tests
- [x] Add documentation with an example

### CODE REVIEW

**Comment by qrohlf** [comment]
This PR looks good to me, and would be helpful for me in my usage of the library. Maintainers, anything holding it up from being merged (other than the merge conflicts)? 

I'm happy to adopt the PR and fix the merge conflicts if there are no other outstanding issues with it.

**Comment by vinnymac** [comment]
@qrohlf I've resolved the merge conflicts.

File: src/Map.js [hunk 1]
```diff
@@ -78,6 +78,18 @@ export class Map extends KeyedCollection {
...
 
+  deleteAll(keys) {
+    var iterable = IndexedIterable(keys);
+
+    if (iterable.size === 0) {
+      return this;
+    }
+
+    return this.withMutations(map => {
+      iterable.forEach((key) => map.remove(key, NOT_SET));
```
Comment (leebyron) [inline src/Map.js]: `.remove()` only accepts one argument

Comment (leebyron) [inline src/Map.js]: please remove the wrapping `()` for consistency with the previous arrow function argument style

File: type-definitions/Immutable.d.ts [hunk 2]
```diff
@@ -601,6 +601,16 @@ declare module Immutable {
...
 
     /**
+     * Returns a new Map which excludes the provided `keys`.
+     *
+     *     var names = Immutable.Map({ a: "Aaron", b: "Barry", c: "Connor" });
+     *     names.deleteAll(['a']); // { b: "Barry", c: "Connor" }
+     *
+     */
+    deleteAll(keys: Array<any>): Map<K, V>;
+    deleteAll(keys: Iterable.Indexed<any>): Map<K, V>;
```
Comment (leebyron) [inline type-definitions/Immutable.d.ts]: `keys: Iterable<any, K>`

File: src/Map.js [hunk 3]
```diff
@@ -78,6 +78,18 @@ export class Map extends KeyedCollection {
     return this.updateIn(keyPath, () => NOT_SET);
   }
 
+  deleteAll(keys) {
```
Comment (leebyron) [inline src/Map.js]: The other delete methods have a `remove` alias. Could you add one here as well

File: src/Map.js [hunk 4]
```diff
@@ -78,6 +78,18 @@ export class Map extends KeyedCollection {
...
+    var iterable = IndexedIterable(keys);
```
Comment (leebyron) [inline src/Map.js]: You should only need `Iterable` - it shouldn't matter that it's indexed

File: type-definitions/Immutable.d.ts [hunk 5]
```diff
@@ -601,6 +601,16 @@ declare module Immutable {
     remove(key: K): Map<K, V>;
 
     /**
+     * Returns a new Map which excludes the provided `keys`.
+     *
+     *     var names = Immutable.Map({ a: "Aaron", b: "Barry", c: "Connor" });
+     *     names.deleteAll(['a']); // { b: "Barry", c: "Connor" }
+     *
+     */
+    deleteAll(keys: Array<any>): Map<K, V>;
```
Comment (leebyron) [inline type-definitions/Immutable.d.ts]: `Array<K>`

**Comment by vinnymac** [comment]
@leebyron I made the requested changes other than implementing `deleteAll` for List.

I attempted to add support for `deleteAll` but `deleteAll([0, -1])` is returning a different result than `.remove(0).remove(-1)` due to my usage of `withMutations`. I am pretty sure #228 is the issue I am seeing.

I can implement this some other ways, but I would like to hear what you would like to do next in this situation before I move forward. Thank you for the advice and your time.

**Comment by leebyron** [comment]
That definitely makes sense thanks for investigating. Of course deletion at indices will change the shape of the List itself. Negative indices cause issue there. Let's start with Map since we know that works. I'll help get a clean merge

**Pull request merged**